framework/Optimizers/SPSA.py

```diff
@@ -326,39 +327,45 @@ def _computeStepSize(self,paramDict,iterNum,traj):
       @ Out, new, float, current value for gain ak
     """
     try:
-      size = self.counter['lastStepSize'][traj]
+      size = self.counter['lastStepSize'][traj][0]
     except KeyError:
       size = paramDict['initialStepSize']
     # modify step size based on the history of the gradients used
     frac = self.fractionalStepChangeFromGradHistory(traj)
     new = size*frac
     self.raiseADebug('step gain size for traj "{}" iternum "{}": {:1.3e} (root {:1.2e} frac {:1.2e})'.format(traj,iterNum,new,size,frac))
-    self.counter['lastStepSize'][traj] = new
+    self.counter['lastStepSize'][traj][0] = new
     return new
 
-  def _computePerturbationDistance(self,traj,paramDict,iterNum):
+  def _computePerturbationDistance(self,traj,paramDict,iterNum,resample=False):
     """
       Utility function to compute the perturbation distance (distance from opt point to grad point)
       @ In, traj, int, integer label for current trajectory
       @ In, paramDict, dict, dictionary containing information to compute gain parameter
       @ In, iterNum, int, current iteration index
+      @ In, resample, bool, optional, if False take the last stepsize as the reference
       @ Out, distance, float, current value for gain ck
     """
     # perturbation point should be a percent of the intended step
     pct = paramDict['pertDist']
-    distance = pct * self.counter['lastStepSize'][traj]
+    if not resample:
+      distance = pct * self.counter['lastStepSize'][traj][0]
+    else:
+      distance = pct * self.counter['lastStepSize'][traj][1]
```

Comment:
Not required, but I would create a new variable `stepSize` that takes either `[0]` or `[1]` from the counter, then have one version of the equation instead of a copy. Minor maintenance improvement.